File: src/main/java/Sigmath/calculus/Calculus.java
package sigmath;

import com.aspose.ms.System.Enum;
import com.aspose.ms.System.Exception;
import com.aspose.ms.System.ArgumentException;
import com.aspose.ms.System.msMath;
import com.aspose.ms.System.Collections.Generic.List;

  
public class Calculus {
    public static double integral(Func<Double, Double> f, double a, double b, int n) {
        double h = (b - a) / n;
        double integral = 0;
        for (int i = 1; i <= n; i++) {
            double x = a + i * h;
            integral += f.apply(x);
        }
        integral += (f.apply(a) + f.apply(b)) / 2;
        integral *= h;
        return integral;
    }
    
    public static double limit(Func<Double, Double> f, Func<Double, Double> g, double x) {
        final int maxIterations = 1000;
        final double tolerance = 1e-9;
        double fx = f.apply(x);
        double gx = g.apply(x);
        double limit = fx / gx;
        for (int i = 0; i < maxIterations; i++) {
            double fx1 = derivative(f, x);
            double gx1 = derivative(g, x);
            double newLimit = fx1 / gx1;
            if (Math.abs(newLimit - limit) < tolerance) {
                return newLimit;
            }
            limit = newLimit;
            x += 1e-9;
        }
        throw new RuntimeException("Limit does not exist or cannot be computed with L'Hopital's Rule.");
    }
    
    public static double findRoot(Func<Double, Double> f, double a, double b, double epsilon) {
        if (f.apply(a) * f.apply(b) > 0)
            throw new IllegalArgumentException("Function must have opposite signs at interval endpoints");
    
        double c = (a + b) / 2;
    
        while (Math.abs(f.apply(c)) > epsilon) {
            if (f.apply(a) * f.apply(c) < 0)
                b = c;
            else
                a = c;
    
            c = (a + b) / 2;
        }
    
        return c;
    }
    
    private static double derivative(Func<Double, Double> f, double x) {
        final double h = 1e-6;
        return (f.apply(x + h) - f.apply(x - h)) / (2 * h);
    }
    
    private static int factorial(int n) {
        int result = 1;
        for (int i = 2; i <= n; i++)
            result *= i;
        return result;
    }
    
    public static Func<Double[], Double> partialDerivative(Func<Double[], Double> f, int index, double h) {
        return (x) -> {
            Double[] xPlus = x.clone();
            xPlus[index] += h;
    
            Double[] xMinus = x.clone();
            xMinus[index] -= h;
    
            return (f.apply(xPlus) - f.apply(xMinus)) / (2 * h);
        };
    }
    
    public static Func<Double[], Double[]> gradient(Func<Double[], Double> f, double h) {
        return (x) -> {
            Double[] gradient = new Double[x.length];
    
            for (int i = 0; i < x.length; i++)
                gradient[i] = partialDerivative(f, i, h).apply(x);
    
            return gradient;
        };
    }
    

    public static Function<double[], double[][]> hessian(Function<double[], Double> f, double h) {
        return x -> {
            int n = x.length;
            double[][] hessian = new double[n][n];

            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    Function<double[], Double> fxx = partialDerivative(partialDerivative(f, i, h), j, h);
                    hessian[i][j] = fxx.apply(x);
                }
            }

            return hessian;
        };
    }
    
    public static double[] findCriticalPoints(Function<Double, Double> f, double a, double b, double epsilon) {
        List<Double> criticalPoints = new ArrayList<>();

        // Evaluate function at endpoints of interval
        double fa = f.apply(a);
        double fb = f.apply(b);

        // Check if endpoints are critical points
        if (Math.abs(fa) < epsilon) {
            criticalPoints.add(a);
        }
        if (Math.abs(fb) < epsilon) {
            criticalPoints.add(b);
        }

        // Find critical points in the interval
        double x = a;
        double fx = f.apply(a);

        while (x < b) {
            double xNext = x + epsilon;
            double fxNext = f.apply(xNext);

            // Check if the current point is a critical point
            if (Math.abs(fx) < epsilon && x > a && x < b) {
                criticalPoints.add(x);
            }

            // Check if the function changes sign between x and xNext
            if (fx * fxNext < 0) {
                // Use bisection method to find a root
                double xLeft = x;
                double xRight = xNext;

                while (xRight - xLeft > epsilon) {
                    double xMid = (xLeft + xRight) / 2;
                    double fxMid = f.apply(xMid);

                    if (Math.abs(fxMid) < epsilon) {
                        criticalPoints.add(xMid);
                        break;
                    }

                    if (fx * fxMid < 0) {
                        xRight = xMid;
                    } else {
                        xLeft = xMid;
                    }
                }
            }

            x = xNext;
            fx = fxNext;
        }

        double[] result = new double[criticalPoints.size()];
        for (int i = 0; i < criticalPoints.size(); i++) {
            result[i] = criticalPoints.get(i);
        }

        return result;
    }
    


    public static /*enum*/ final class RiemannSumType extends Enum
    {
	private RiemannSumType(){}	
        public static final int Left = 0;
        public static final int Right = 1;
        public static final int Midpoint = 2;
        public static final int Trapezoidal = 3;

	static {
		Enum.register(new Enum.SimpleEnum(RiemannSumType.class, Integer.class) {{
		addConstant("Left", Left);
		addConstant("Right", Right);
		addConstant("Midpoint", Midpoint);
		addConstant("Trapezoidal", Trapezoidal);
		}});
	}}

    public static double riemannSum(Function<Double, Double> f, double a, double b, int n, RiemannSumType type)
    {
        double dx = (b - a) / n;
        double sum = 0;
    
        for (int i = 0; i < n; i++)
        {
            double x = a + i * dx;
            double fx = f.apply(x);
    
            switch (type)
            {
                case Left:
                    sum += fx * dx;
                    break;
                case Right:
                    sum += f.apply(x + dx) * dx;
                    break;
                case Midpoint:
                    sum += f.apply(x + dx / 2) * dx;
                    break;
                case Trapezoidal:
                    sum += (fx + f.apply(x + dx)) * dx / 2;
                    break;
            }
        }
    
        return sum;
    }
    
    public static double definiteIntegral(Function<Double, Double> f, double a, double b, int n)
    {
        return riemannSum(f, a, b, n, RiemannSumType.Midpoint);
    }
    
    public static double definiteIntegral(Function<Double, Double> f, double a, double b)
    {
        return riemannSum(f, a, b, 1000, RiemannSumType.Midpoint);
    }
}

File: Sigmath/Sigmath/Sigmath.ConvertedToVBNet/calculus/calculus.vb
Imports System.Collections.Generic

Public Class Calculus
	Public Shared Function Derivative(f As Func(Of Double, Double), x As Double) As Double
		Const  dx As Double = 1E-09
		' small increment to find the derivative
		Dim y1 As Double = f(x)
		Dim y2 As Double = f(x + dx)
		Dim derivative__1 As Double = (y2 - y1) / dx
		Return derivative__1
	End Function

	Public Shared Function Integral(f As Func(Of Double, Double), a As Double, b As Double, n As Integer) As Double
		Dim h As Double = (b - a) / n
		Dim integral__1 As Double = 0
		For i As Integer = 1 To n - 1
			Dim x As Double = a + i * h
			integral__1 += f(x)
		Next
		integral__1 += (f(a) + f(b)) / 2
		integral__1 *= h
		Return integral__1
	End Function

	Public Shared Function Limit(f As Func(Of Double, Double), g As Func(Of Double, Double), x As Double) As Double
		Const  maxIterations As Integer = 1000
		Const  tolerance As Double = 1E-09
		Dim fx As Double = f(x)
		Dim gx As Double = g(x)
		Dim limit__1 As Double = fx / gx
		For i As Integer = 0 To maxIterations - 1
			Dim fx1 As Double = Derivative(f, x)
			Dim gx1 As Double = Derivative(g, x)
			Dim newLimit As Double = fx1 / gx1
			If Basic.Abs(newLimit - limit__1) < tolerance Then
				Return newLimit
			End If
			limit__1 = newLimit
			x += 1E-09
		Next
		Throw New Exception("Limit does not exist or cannot be computed with L'Hopital's Rule.")
	End Function

	Public Shared Function FindRoot(f As Func(Of Double, Double), a As Double, b As Double, Optional epsilon As Double = 1E-06) As Double
		If f(a) * f(b) > 0 Then
			Throw New ArgumentException("Function must have opposite signs at interval endpoints")
		End If

		Dim c As Double = (a + b) / 2

		While Math.Abs(f(c)) > epsilon
			If f(a) * f(c) < 0 Then
				b = c
			Else
				a = c
			End If

			c = (a + b) / 2
		End While

		Return c
	End Function

	Private Shared Function Factorial(n As Integer) As Integer
		Dim result As Integer = 1

		For i As Integer = 2 To n
			result *= i
		Next

		Return result
	End Function

	Public Shared Function PartialDerivative(f As Func(Of Double(), Double), index As Integer, Optional h As Double = 1E-06) As Func(Of Double(), Double)
		Return Function(x) 
		Dim xPlus = DirectCast(x.Clone(), Double())
		xPlus(index) += h

		Dim xMinus = DirectCast(x.Clone(), Double())
		xMinus(index) -= h

		Return (f(xPlus) - f(xMinus)) / (2 * h)

End Function
	End Function

	Public Shared Function Gradient(f As Func(Of Double(), Double), Optional h As Double = 1E-06) As Func(Of Double(), Double())
		Return Function(x) 
		Dim gradient__1 = New Double(x.Length - 1) {}

		For i As Integer = 0 To x.Length - 1
			gradient__1(i) = PartialDerivative(f, i, h)(x)
		Next

		Return gradient__1

End Function
	End Function

	Public Shared Function Hessian(f As Func(Of Double(), Double), Optional h As Double = 1E-06) As Func(Of Double(), Double(,))
		Return Function(x) 
		Dim n As Integer = x.Length
		Dim hessian__1 = New Double(n - 1, n - 1) {}

		For i As Integer = 0 To n - 1
			For j As Integer = 0 To n - 1
				Dim fxx = PartialDerivative(PartialDerivative(f, i, h), j, h)(x)
				hessian__1(i, j) = fxx
			Next
		Next

		Return hessian__1

End Function
	End Function

	Public Shared Function FindCriticalPoints(f As Func(Of Double, Double), a As Double, b As Double, Optional epsilon As Double = 1E-06) As Double()
		Dim criticalPoints = New List(Of Double)()

		' Evaluate function at endpoints of interval
		Dim fa As Double = f(a)
		Dim fb As Double = f(b)

		' Check if endpoints are critical points
		If Math.Abs(fa) < epsilon Then
			criticalPoints.Add(a)
		End If
		If Math.Abs(fb) < epsilon Then
			criticalPoints.Add(b)
		End If

		' Find critical points in the interval
		Dim x As Double = a
		Dim fx As Double = f(a)

		While x < b
			Dim xNext As Double = x + epsilon
			Dim fxNext As Double = f(xNext)

			' Check if the current point is a critical point
			If Math.Abs(fx) < epsilon AndAlso x > a AndAlso x < b Then
				criticalPoints.Add(x)
			End If

			' Check if the function changes sign between x and xNext
			If fx * fxNext < 0 Then
				' Use bisection method to find a root
				Dim xLeft As Double = x
				Dim xRight As Double = xNext

				While xRight - xLeft > epsilon
					Dim xMid As Double = (xLeft + xRight) / 2
					Dim fxMid As Double = f(xMid)

					If Math.Abs(fxMid) < epsilon Then
						criticalPoints.Add(xMid)
						Exit While
					End If

					If fx * fxMid < 0 Then
						xRight = xMid
					Else
						xLeft = xMid
					End If
				End While
			End If

			x = xNext
			fx = fxNext
		End While

		Return criticalPoints.ToArray()
	End Function

	Public Enum RiemannSumType
		Left
		Right
		Midpoint
		Trapezoidal
	End Enum

	Public Shared Function RiemannSum(f As Func(Of Double, Double), a As Double, b As Double, n As Integer, Optional type As RiemannSumType = RiemannSumType.Left) As Double
		Dim dx As Double = (b - a) / n
		Dim sum As Double = 0

		For i As Integer = 0 To n - 1
			Dim x As Double = a + i * dx
			Dim fx As Double = f(x)

			Select Case type
				Case RiemannSumType.Left
					sum += fx * dx
					Exit Select
				Case RiemannSumType.Right
					sum += f(x + dx) * dx
					Exit Select
				Case RiemannSumType.Midpoint
					sum += f(x + dx / 2) * dx
					Exit Select
				Case RiemannSumType.Trapezoidal
					sum += (fx + f(x + dx)) * dx / 2
					Exit Select
			End Select
		Next

		Return sum
	End Function

	Public Shared Function DefiniteIntegral(f As Func(Of Double, Double), a As Double, b As Double, Optional n As Integer = 1000) As Double
		Return RiemannSum(f, a, b, n, RiemannSumType.Midpoint)
	End Function
End Class
